hamilton/lifecycle/caching.py

```diff
@@ -0,0 +1,1087 @@
+        """Execution logs of the cache adapter.
+
+        :param run_id: id of a previously logged run. If None, return all logged runs
+        :return: a mapping between node/task and a list of logged events
+        """
+        if run_id:
+            run_logs = collections.defaultdict(list)
+            for event in self._logs[run_id]:
+                key = (event.node_name, event.task_id) if event.task_id else event.node_name
+                run_logs[key].append(event)
+            return {k: v for k, v in run_logs.items()}
+        return self._logs
+
+    def get_memory_metadata(self, node_name: str, task_id: Optional[str] = None) -> Optional[str]:
+        """Get in-memory data_version whether a task_id is specified or not.
+        Returns ``None`` if no data_version is found
+        """
+        # default behavior; when not using task-based execution
+        if task_id is None:
+            data_version = self.data_versions.get(node_name, SENTINEL)
+        # nodes outside are uniquely identifiable by node_name
+        elif node_name == task_id:
+            data_version = self.data_versions.get(node_name, SENTINEL)
+        # because `Parallelizable` node return the collected generator as a result
+        # this is not the data we want to version.
+        elif task_id == f"expand-{node_name}":
+            data_version = SENTINEL
+        elif "collect-" in task_id:
+            data_version = self.data_versions.get(node_name, SENTINEL)
+        else:
+            data_version = self.data_versions.get(node_name, {}).get(task_id, SENTINEL)  # type: ignore ; `task_id` can't be None
+
+        if data_version is SENTINEL:
+            self._log_event(
+                node_name=node_name,
+                task_id=task_id,
+                actor="adapter",
+                event_type=CachingEventType.GET_DATA_VERSION,
+                msg="miss",
+            )
+            return None
+        else:
+            self._log_event(
+                node_name=node_name,
+                task_id=task_id,
+                actor="adapter",
+                event_type=CachingEventType.GET_DATA_VERSION,
+                msg="hit",
+                value=data_version,
+            )
+            return data_version
+
+    def set_memory_metadata(
+        self, node_name: str, data_version: str, task_id: Optional[str] = None
+    ) -> None:
+        """Set in-memory data_version whether a task_id is specified or not"""
+        assert data_version is not None
+        # default behavior; when not using task-based execution
+        if task_id is None:
+            self.data_versions[node_name] = data_version
+        # nodes outside are uniquely identifiable by node_name
+        elif node_name == task_id:
+            self.data_versions[node_name] = data_version
+        # nodes "on the edge" of parallelizable/collect blocks are uniquely identifiable by node_name
+        elif task_id == f"expand-{node_name}":
```

Comment:
Definitely a pain to manage, but laying out all cases with `if/elif/else` makes for more readable and easier to maintain code. Hooks didn't provide more reliable information other than the `task_id`.

The cases are also pretty explicitly described in the comments. Typical cases:

- `task_id is None`: there's no `Parallelizable/Collect` used
- `task_id == node_name`: this is outside of a block 
- `task_id == f"expand-{node_name}`: this is a `Parallelizable` node
- `"collect-" in task_id`: this is a `Collect` node. The check is different than for `Parallelizable` because the task_id contains the name of the `Parallelizable` node, not the "current" one.
- else (i.e., `task_id != node_name`) this is inside a parallel/collect block

The best solution to avoid these would be to parse the `HamiltonGraph` in `run_before_graph_execution` and assign an enum "inside, outside, etc.", but this is redundant with the `node.NodeType` that's available in node-related hooks